src/QueryObserver.ts

```diff
@@ -1,7 +1,101 @@
+import { Action, Query, QueryFunction } from "./Query"
 import { Subscribable } from "./subscribable"
+import { shallowEqualObjects } from "./utils"
 
-export class QueryObserver extends Subscribable {
-  onQueryUpdate(data: unknown): void {
-    console.log(data)
+export type QueryObserverListener<TData, TError> = (
+  result: QueryResult<TData, TError>
+) => void
+
+export interface QueryResult<TData, TError> {
+  data: TData | undefined
+  error: TError | null
+  status: "idle" | "error" | "loading" | "success"
+  isLoading: boolean
+  isError: boolean
+  isSuccess: boolean
+}
+
+export interface QueryObserverOptions<TData, TError> {
+  fetchOnMount?: boolean
+  listeners: QueryObserverListener<TData, TError>[]
+}
+
+const INITIAL_QUERY_OBSERVER_RESULT = {
+  data: undefined,
+  error: null,
+  status: "idle",
+  isLoading: false,
+  isError: false,
+  isSuccess: false,
+}
+
+export class QueryObserver<
```

Comment:
Fleshed out the `QueryObserver` class to have an idea of how the `core` API would look like. I didn't really do much async fetching with anything else than `$.ajax` from `jQuery` in my Vanilla JS days, so you might have a better idea for an API than I've had. 😄 

In the end, I think the nice thing about this is that, when translating this into a react context, we can just `.subscribe()` listener for the state updates in the `QueryObserver` and update the hook state accordingly. Of course, there's stuff to think about, such as making sure this doesn't get unmounted, etc...